libdnf/rpm/solv/map.hpp

```diff
@@ -0,0 +1,199 @@
+    SolvMap(const SolvMap & other);
+
+    ~SolvMap();
+
+    // GENERIC OPERATIONS
+
+    /// Return the underlying libsolv Map
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.getMap()
+    //Map * getMap() const;
+    //Map *PackageSet::getMap() const { return &pImpl->map; }
+    const Map * get_map() const;
+
+    /// Return the number of solvables in the SolvMap (number of 1s in the bitmap).
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.size()
+    std::size_t size() const;
+
+    bool empty() const;
+
+    void clear();
+
+    // ITEM OPERATIONS
+
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.set(Id id)
+    void add(Id id);
+
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.has(Id id)
+    bool contains(Id id) const;
+
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.remove(Id id)
+    void remove(Id id);
+
+    // SET OPERATIONS - Map
+
+    /// Union operator
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.operator+=(const Map * other)
+    SolvMap & operator|=(const Map * other);
+
+    /// Difference operator
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.operator-=(const Map * other)
+    SolvMap & operator-=(const Map * other);
+
+    /// Intersection operator
+    ///
+    /// @replaces libdnf:sack/packageset.hpp:method:PackageSet.operator/=(const Map * other)
+    SolvMap & operator&=(const Map * other);
+
+    // SET OPERATIONS - SolvMap
+
+    /// Union operator
+    SolvMap & operator|=(const SolvMap & other);
+
+    /// Difference operator
+    SolvMap & operator-=(const SolvMap & other);
+
+    /// Intersection operator
+    SolvMap & operator&=(const SolvMap & other);
+
+protected:
+    /// Check if `id` is in bitmap range.
+    /// Throws std::out_of_range
+    void check_id_in_bitmap_range(Id id) const;
+
+    /// Check if the `other` bitmap has the same size as `map`.
+    void check_equal_bitmap_sizes(const Map * other) const;
+
+    // ITEM OPERATIONS - UNSAFE
+
+    /// Faster, but unsafe version of add() method that is doesn't check bitmap range
+    void add_unsafe(Id id);
+
+    /// Faster, but unsafe version of contains() method that is doesn't check bitmap range
+    bool contains_unsafe(Id id) const;
+
+    /// Faster, but unsafe version of remove() method that is doesn't check bitmap range
+    void remove_unsafe(Id id);
+
+private:
+    Map map;
+};
+
+
+inline void SolvMap::check_id_in_bitmap_range(Id id) const {
+    if (id < 0 || id >= (map.size << 3)) {
+        throw std::out_of_range("Id is out of bitmap range");
+    }
+}
+
+
+inline void SolvMap::check_equal_bitmap_sizes(const Map * other) const {
+    if (map.size != other->size) {
+        throw std::out_of_range("Bitmap sizes are not equal");
+    }
+}
+
+
+inline void SolvMap::add(Id id) {
+    check_id_in_bitmap_range(id);
+    add_unsafe(id);
+}
+
+
+inline bool SolvMap::contains(Id id) const {
```

Comment:
I know that "safety" of the method was required by @j-mracek. But what is "safe" implementation of the `contains` method in the set. I usually expect result `true` or `false`. The item is or is not present in the set.
What means exceptions out_of_range?  In fact it says the item is not present in the set. So `false` is the right result. 
I know how it is implemented internally but for user is the third state - exception - strange.

Reply:
This can be seen from a different perspective.
The underlying bitmap represents all allowed values.
You shouldn't be working with any other values and it shouldn't be possible with Packages under normal circumstances. For me, using an Id outside the bitmap range is clearly an error condition and we should be informed. This may happen if a bitmap is extended in one place, but we forget to do that in another place. Then returning false would be clearly a mistake.

Users shouldn't see this exception, because they will work with Sack and Package objects built on top of bitmap and the Packages should *always* exist in the bitmap.

Reply:
Changed to return false when Id is out of bitmap range.